File: go/leet_code/1544.go
// Package leet_code provides solutions for LeetCode problems.
//
// This file provides a solution for the problem of making a string good.
// A string is good if it does not have two adjacent characters that are the same, but one is uppercase and the other is lowercase.
package leet_code

// makeGood makes a string good.
//
// It accepts one parameter:
// - s: a string that only contains lowercase and uppercase English letters.
//
// The function returns a string that is a good string and is a subsequence of s.
//
// The function uses a stack to keep track of the characters in the string.
// It iterates over the string, and for each character, if the stack is not empty and the top character of the stack is the same as the current character but with different case, it pops the top character from the stack, otherwise, it pushes the current character to the stack.
//
// Time complexity analysis:
// - Best-case: O(n), when the string is already good.
// - Worst-case: O(n), when all characters are the same but with alternating cases.
// - Average-case: O(n), as we always have to iterate over each character in the string.
func makeGood(s string) string {
	stack := make([]byte, 0)
	for i := 0; i < len(s); i++ {
		if len(stack) == 0 {
			stack = append(stack, s[i])
			continue
		}
		if stack[len(stack)-1] == s[i]+32 || stack[len(stack)-1] == s[i]-32 {
			stack = stack[:len(stack)-1]
		} else {
			stack = append(stack, s[i])
		}
	}
	return string(stack)
}

File: dart/leet_code/1544.dart
/// Dart implementation of the LeetCode problem 1544: Make The String Great.
///
/// This module contains a class `Solution` with a method `makeGood`.
/// The `makeGood` method takes in one parameter:
/// - `s`: A string representing the input string.
///
/// The method returns a string representing the "good" string after making the necessary modifications.
/// A string is "good" if there are no two adjacent characters that are of the same letter but different case.
///
/// Error Handling:
/// - This method assumes that the input parameter is well-formed, i.e., `s` is a string of letters.
/// - If the input parameter is not well-formed, the behavior of the method is undefined.
library;

class Solution {
  /// Makes the input string "good".
  ///
  /// This method uses a stack to keep track of the characters in the string.
  /// It iterates over each character in the input string. If the stack is not empty and the last character in the stack is the same letter but different case as the current character, it removes the last character from the stack.
  /// Otherwise, it adds the current character to the stack.
  /// After the iteration, it joins the characters in the stack to form the "good" string.
  ///
  /// @param s The input string.
  /// @return A string representing the "good" string.
  String makeGood(String s) {
    List<String> stack = [];
    for (int i = 0; i < s.length; i++) {
      if (stack.isNotEmpty && stack.last.toLowerCase() == s[i].toLowerCase() && stack.last != s[i]) {
        stack.removeLast();
      } else {
        stack.add(s[i]);
      }
    }
    return stack.join('');
  }
}
